### PR TITLE
Updating the linthresh guessing

### PR DESCRIPTION
**Pull request by chrishavlin**
This is an attempt at updating how the linthresh value is calculated for SymLogNorms when not provided. Note that I based it off of #3849 , so I'm submitting in draft mode until that one is cleared. The changes here are simple enough that it should be easy to update this PR after #3849 is fully reviewed. 

IMO the linthresh calculation here nicely handles all of the concerns of #3944 without too much complexity. I need to update/add some tests and the docs would need a little updating, but it'd be great to get some feedback on the approach. 

Conceptually, the calculation here uses the minimum absolute nonzero value as a starting guess at linthresh (as in  #3295) but then limits the dynamic range of the symlog scale when necessary to avoid the floating point precision issues that caused the errors in #3858. The limiting is sensitive to the data so it does a decent job for a range of data distributions (and should not error for any).  

Some images:

```python
fn = 'FIRE_M12i_ref11/snapshot_600.hdf5'
ds = yt.load(fn)
p = yt.ProjectionPlot(ds, 'x', ('gas', 'density'), width=(30, 'Mpc'))
p.show()
```
image

And using the sample data from #3858 
```python
import yt

ds = yt.load('symlogsets/yt_bug_reproducer/plt04000')
slc = yt.SlicePlot(ds, "z", ("boxlib", "scalar"))
slc.show()
```

image

### CODE REVIEW

**Comment by chrishavlin** [comment]
Note that I experimented with the new imshow kwarg (`interpolation_stage="rgba"`) but still got some rounding error crashes related to the normalization objects (`SymLogNorm`, `LogNorm`) and so gave up on that route...

File: yt/visualization/_handlers.py [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+        actual_log_range = np.log10(max_abs_val) - np.log10(min_abs_val)
+        if actual_log_range < cutoff_sigdigs:
+            cutoff_sigdigs = np.ceil(actual_log_range)
+
+        if np.log10(max_abs_val) - np.log10(linthresh) > cutoff_sigdigs:
+            # this is required to avoid rounding error in the matplotlib normalization
+            # when the range is too large.
+            linthresh = max_abs_val / (10**cutoff_sigdigs)
+
+        return linthresh
```
Comment (chrishavlin) [inline yt/visualization/_handlers.py:379]: So this section is the relevant change...

File: yt/visualization/_handlers.py [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+    def _guess_linthresh(self, data):
+        # data is the ImageArray to plot
+
+        # get the data range as unyt quantities
+        nz_data = data[data != 0]
+        min_abs_val, max_abs_val = np.sort(
+            (np.abs(np.nanmin(nz_data)), np.abs(np.nanmax(nz_data)))
+        )
+        min_abs_val = self.ds.quan(min_abs_val, data.units)
+        max_abs_val = self.ds.quan(max_abs_val, data.units)
```
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/_handlers.py:361]: ⚠️ Actually `data` cannot be assumed to even *have* units... (see line 317 for instance)

**Review by neutrinoceros — COMMENTED** [review]
Thank you so much for continuing this gianormous task !
I think I spotted a couple issues with the current approach, but nothing that we can't address.
A related question: did you decide that using `interpolation_stage` (as discussed in #3944) wasn't a reliable solution in the end, and why ?

File: yt/visualization/_handlers.py [hunk 3]
```diff
@@ -327,45 +327,19 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
             # allowing to toggle between lin and log scaling without detailed user input
             norm_type = self.norm_type
         else:
-            if kw["vmin"] == kw["vmax"] or not np.any(np.isfinite(data)):
+            if kw["vmin"] == kw["vmax"] or not np.any(finite_values_mask):
                 norm_type = Normalize
             elif kw["vmin"] <= 0:
                 norm_type = SymLogNorm
-            elif (
-                Version("3.3") <= MPL_VERSION < Version("3.5")
```
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py]: Two reasons: first is that the preceeding `kw["vmin"] <= 0"` condition would always catch before this one (if `kw['vmin']==0` then the preceeding `kw['vmin'] <= 0` will always be True) . I considered flagging it in your PR but since it's closely related to symlog/linthresh settings it seemed better to address it here. 

The second reason I went for removal instead of changing the condition is that controlling for the issue when the range is too large happens down in the linthresh calculation so this seemed redundant. Thinking about it now though, I do think it would actually make sense to allow `LogScale`  when possible... but I think this condition probably needs some adjusting rather than straight removal.

I also mean to test this more thoroughly will different matplotlib versions before switching from draft -- so far just been using the latest release (3.5.2).

File: yt/visualization/_handlers.py [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+        max_abs_val = self.ds.quan(max_abs_val, data.units)
+        min_abs_nonzero = self.ds.quan(np.nanmin(np.abs(nz_data)), data.units)
+
+        linthresh = min_abs_nonzero
+        # the above works well for most data. But need to control for a few
+        # edge cases related to floating point precision errors.
+
+        # first, set the scaling factor
+        cutoff_sigdigs = 30.0  # max allowable range in significant digits
+        actual_log_range = np.log10(max_abs_val) - np.log10(min_abs_val)
```
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py:370]: oh, good call. didn't think about mixed signs possibly having the same abs value. If there's not one yet, I'll add a test for that special case too. 

Ya, renaming variables for clarity throughout is a good call.

File: yt/visualization/_handlers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py:361]: yikes, good catch. will add a similar attribute check for units.

File: yt/visualization/_handlers.py [hunk 5]
```diff
@@ -327,45 +327,19 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
-            min_abs_val, max_abs_val = np.sort(
-                np.abs((self.to_float(np.nanmin(data)), self.to_float(np.nanmax(data))))
-            )
             if self.linthresh is not None:
                 linthresh = self.to_float(self.linthresh)
-            elif min_abs_val > 0:
-                linthresh = min_abs_val
             else:
-                linthresh = max_abs_val / 1000
+                linthresh = self.to_float(self._guess_linthresh(data))
```
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py:341]: ya, could definitely return the float value from `_guess_linthresh`. I initially had `_guess_linthresh` as a standalone function so put the to_float here and forgot to move it to `_guess_linthresh` after I made it a class method :) will do that!

**Comment by chrishavlin** [comment]
So the issue with using `interpolation_stage` is that it is only an `imshow` kwarg. When using the `Norm` objects directly, it's pretty easy to get it to error with the edge cases we're trying to handle here. e.g., using a linthresh close to the underflow limit will error:

```
from matplotlib.colors import SymLogNorm
v = SymLogNorm(1e-323, vmin=-1e-1, vmax=1e-1)
v(1e-2)  # errors with the familiar `ValueError: Invalid vmin or vmax`. 
```
There's some discussion of adding an interpolation stage argument to the `Norm` objects (e.g, the end of this [issue](https://github.com/matplotlib/matplotlib/issues/21167)), but even if that happens soon we'd then be handling multiple dependency conditions for the same underlying problem. And so I feel like it's a lot simpler to just try to avoid the conditions that give rise to the precision issues.

I mentioned this in one of the replies, in addition to adding to the existing yt tests, I'm planning to do a more thorough matplotlib-dependent suite of tests to check all these edge cases before switching from draft.

**Comment by neutrinoceros** [comment]
Seems reasonable, thank you for your research on this !

File: yt/visualization/_handlers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (neutrinoceros) [reply yt/visualization/_handlers.py]: okay sounds good !

**Comment by chrishavlin** [comment]
Ok, so I rewrote all of `_guess_linthresh`. It felt clearer to avoid the whole abs(min,max) vs min(abs) confusion and treat the positive and negative ranges separately. It's maybe a little more verbose but I think it's easier to understand now.

File: yt/visualization/_handlers.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py]: So I'm back to thinking I'd rather remove these lines... they're not needed in that sense that removing them does not result in errors. But it is true we'd be switching to SymLogNorm in situations that could technically use a LogNorm -- If we wanted to, we could allow a LogNorm when the data is >= 0 if matplotlib >= 3.5.2 **and** if we supply that new interpolation_stage keyword to imshow. So happy to implement that here if you think it's a good idea, but i'd rather wait til your #3849 is merged before doing so.

File: yt/visualization/_handlers.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrishavlin) [reply yt/visualization/_handlers.py]: _guess_linthresh will now always return a plain float

**Comment by chrishavlin** [comment]
interesting... that failed test passed locally. Will have to dig in further...

**Comment by neutrinoceros** [comment]
#3849 is in, now this should be rebased so we can see only the relevant bits in the diff

**Comment by neutrinoceros** [comment]
I went for it since the rebase turned out a no op.

File: yt/visualization/_handlers.py (hunk 1 above)
Comment (chummels) [reply yt/visualization/_handlers.py]: This is a smart algorithm.  It seems to match my solution from #3295 but also incorporate the fix for the sigdig error that my solution broke.  Nice way to find a solution!

**Comment by chummels** [comment]
I like this solution overall.  The only thing I see missing from the PR is a brief mention of this in the colorbar linthresh narrative docs, like a sentence describing the linthresh guess (e.g., "linthresh uses the minimum absolute nonzero value but limits the dynamic range of the symlog scale when necessary to avoid floating point precision issues (10**15)").  Thanks for coding this up, Chris!

**Review by neutrinoceros — COMMENTED** [review]
+1 for documentation. Specifically, the following paragraph should be updated
https://github.com/yt-project/yt/blob/dc7da9e8d56f583656361603275d3f9b55061f1d/doc/source/visualizing/plots.rst#L1000-L1005

I also note that the only 2 test failures we see now are 100% expected and desirable consequences of the change, so we can bump baselines here.

File: yt/visualization/_handlers.py [hunk 6]
```diff
@@ -326,45 +326,24 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
         else:
-            if kw["vmin"] == kw["vmax"] or not np.any(np.isfinite(data)):
+            if kw["vmin"] == kw["vmax"] or not np.any(finite_values_mask):
                 norm_type = Normalize
             elif kw["vmin"] <= 0:
+                # note: matplotlib >= 3.5.2 could use LogNorm here if vmin==0
+                # if we do the following: set vmin to the min nonzero value and
+                # supply the interpolation_stage='rgba' keyword arg to imshow.
+                # this would, however, change default behavior and break answer
+                # tests, so needs a wider discussion.
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:336]: The expected gain is not clear to me from just this comment.
If you think a discussion is still needed after this PR is merged, we could keep #3944 open for discussion,
but I'm not sure I agree of this being a temporary comment with no clear expiration condition.

File: yt/visualization/_handlers.py [hunk 7]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+        pos_min, pos_max = get_minmax(finite_plot_data[finite_plot_data > 0])
+        neg_min, neg_max = get_minmax(finite_plot_data[finite_plot_data < 0])
+
+        has_pos = pos_min is not None
+        has_neg = neg_min is not None
+
+        # the starting guess is the absolute value of the point closest to 0
+        # (remember: neg_max is closer to 0 than neg_min)
+        if has_pos and has_neg:
+            linthresh = np.min((np.abs(neg_max), pos_min))
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:376]: Unless you think it's less clear (builtin `abs` would also work btw)
```suggestion
            linthresh = np.min((-neg_max, pos_min))
```

File: yt/visualization/_handlers.py [hunk 8]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+        elif has_pos:
+            linthresh = pos_min
+        elif has_neg:
+            linthresh = np.abs(neg_max)
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:380]: Same as above
```suggestion
            linthresh = -neg_max
```

File: yt/visualization/_handlers.py [hunk 9]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+        else:
+            # this condition should handled before here in get_norm
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:382]: I would probably not work out of the box here, but FTR this looks like a place where `typing.assert_never` would be useful. (Python 3.11+ only)
```suggestion
            # this condition should be handled before here in get_norm
```

File: yt/visualization/_handlers.py [hunk 10]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+            raise RuntimeError("No negative or positive values")
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:383]: Isn't the only way to end up here that there are no finite data points ?

File: yt/visualization/_handlers.py [hunk 11]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+
+        log10_linthresh = np.log10(linthresh)
+
+        # if either the pos or neg ranges exceed cutoff_sigdigs, then
+        # linthresh is shifted to decrease the range to avoid floating point
+        # precision errors in the normalization.
+        cutoff_sigdigs = 15.0  # max allowable range in significant digits
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:390]: An integer feels more appropriate
```suggestion
        cutoff_sigdigs = 15  # max allowable range in significant digits
```

File: yt/visualization/_handlers.py [hunk 12]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+        if has_pos and np.log10(pos_max) - log10_linthresh > cutoff_sigdigs:
+            linthresh = pos_max / (10.0**cutoff_sigdigs)
+            log10_linthresh = np.log10(linthresh)
+
+        if has_neg and np.log10(np.abs(neg_min)) - log10_linthresh > cutoff_sigdigs:
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:395]: Same as above
```suggestion
        if has_neg and np.log10(-neg_min) - log10_linthresh > cutoff_sigdigs:
```

File: yt/visualization/_handlers.py [hunk 13]
```diff
@@ -374,6 +353,54 @@ def get_norm(self, data: np.ndarray, *args, **kw) -> Normalize:
...
+            linthresh = np.abs(neg_min) / (10.0**cutoff_sigdigs)
+
+        if hasattr(linthresh, "units"):
```
Comment (neutrinoceros) [inline yt/visualization/_handlers.py:398]: Unlikely that as astropy units array could ever make it to this line, and if it does, we don't want to let it pass, do we ?
```suggestion
        if isinstance(linthresh, unyt_quantity):
```

File: yt/visualization/tests/test_plotwindow.py [hunk 14]
```diff
@@ -768,15 +768,57 @@ def test_symlog_min_zero():
...
+    arr2 = np.full(shape, 5.0e-324)
+    arr2[0, 0] = -1e12
+
+    arr3 = arr.copy()
+    arr3[4, 4] = 0.0
+
+    d = {"scalar_spans_0": arr, "tiny_vmax": arr2, "scalar_tiny_with_0": arr3}
+
+    ds = load_uniform_grid(d, shape)
+    for field in d.keys():
```
Comment (neutrinoceros) [inline yt/visualization/tests/test_plotwindow.py:786]:
```suggestion
    for field in d:
```

File: yt/visualization/tests/test_plotwindow.py [hunk 15]
```diff
@@ -768,15 +768,57 @@ def test_symlog_min_zero():
...
+    arr = np.full(shape, -1e30)
+    arr[1, 1] = -1e27
+    arr[2, 2] = 1e-12
+    arr[3, 3] = 1e-10
+
+    arr2 = -1 * arr.copy()  # also check the reverse
+    d = {"linthresh_gt_vmax": arr, "linthresh_lt_vmin": arr2}
+
+    ds = load_uniform_grid(d, shape)
+    for field in d.keys():
```
Comment (neutrinoceros) [inline yt/visualization/tests/test_plotwindow.py:805]:
```suggestion
    for field in d:
```

**Review by neutrinoceros — COMMENTED** [review]
this is excellent. I got a couple suggestions, mostly nits.

File: yt/visualization/_handlers.py (hunk 6 above)
Comment (chrishavlin) [reply yt/visualization/_handlers.py:336]: No need to keep 3944 open -- I'll rephrase this comment to be more of a historical note rather than a "to do" (which is how it currently sounds).

File: yt/visualization/_handlers.py (hunk 7 above)
Comment (chrishavlin) [reply yt/visualization/_handlers.py:376]: I actually think it's a little **more** clear with the `-`! Will change it.

File: yt/visualization/_handlers.py (hunk 10 above)
Comment (chrishavlin) [reply yt/visualization/_handlers.py:383]: oh, yes. I can switch this to read `no finite data points`.

**Comment by neutrinoceros** [comment]
@chrishavlin there's a missing import reported by flake8

**Comment by chrishavlin** [comment]
Latest batch updates the docs, bumps the answers and incorporates the remaining suggestions. I'll go through and close out obviously outdated threads...

File: yt/visualization/_handlers.py (hunk 1 above)
Comment (chrishavlin) [reply yt/visualization/_handlers.py]: The code snippet rendered above is out of date. The re-written version is the same end result (matching 3295 except in the edge cases), but I re-wrote it to handle the negative and positive ranges sequentially cause it made handling edge cases easier. Just wanted to double check that you saw that!

**Review by neutrinoceros — APPROVED** [review]
LGTM. I'll give @chummels a couple days in case he wants to review this.

**Review by chummels — APPROVED** [review]
Looks good!  Thanks!
